### ISSUE TITLE
[BUG] model.predict() is providing random top-k recommendations for RetrievalModel

### ISSUE THREAD

**Issue by gabrielspmoreira** [NVIDIA-Merlin/models#287]
### Bug description
The `model.predict()` method is providing random top-k recommendations for `RetrievalModel`.

### How to reproduce
1 - Download the [sample dataset](https://drive.google.com/file/d/10LqUW768FxhaVL4rx8lVJar_Mhbx-MvI/view?usp=sharing)
2 - Run the following code as it is. It will print `2 - EVALUATION METRICS` with `'recall_at_1000': 0.1835` and also print a very close RECALL@1000 for `4 - MANUAL TOP-K PREDICTION`, which is expected.
3 - Uncomment the `BLOCK-01` in the code. The `2 - EVALUATION METRICS`  will be the same, but now you will now notice that for `4 - MANUAL TOP-K PREDICTION` the `RECALL@1000` will be close to random, as shuffling predictions gives a similar metric result (~ `0.019`). So it is weird that calling `topk_recommender_model.predict()` affects MANUAL TOP-K PREDICTION results (which does not use the `TopKIndexBlock`). This fact might help to isolate better what might be generating the issue.
4 - Uncomment the `BLOCK-02` in the code. Now metrics are computed also for `3 - TOP-K PREDICTION` using the Topk_recommender and `RECALL@1000` will be close to random predictions (~ `0.019`) for both `3 - TOP-K PREDICTION` and for `4 - MANUAL TOP-K PREDICTION` .

### Code to reproduce
```python
import tensorflow as tf

tf.keras.utils.set_random_seed(52)

import cupy
import merlin.models.tf as mm
import merlin.models.tf.dataset as tf_dataloader
import numpy as np
from merlin.io.dataset import Dataset
from merlin.models.tf.blocks.core.index import IndexBlock
from merlin.models.utils.dataset import unique_rows_by_features
from merlin.schema import Schema
from merlin.schema.io.tensorflow_metadata import TensorflowMetadata
from merlin.schema.tags import Tags
from nvtabular.ops import *
from tensorflow.keras import regularizers

data_path = "./data/processed"


schema = TensorflowMetadata.from_proto_text_file(
    data_path + "/train/"
).to_merlin_schema()


schema = schema.select_by_tag([Tags.ITEM_ID, Tags.USER_ID, Tags.ITEM, Tags.USER])

model = mm.TwoTowerModel(
    schema,
    query_tower=mm.MLPBlock(
        [128, 64],
        no_activation_last_layer=True,
        kernel_regularizer=regularizers.l2(1e-2),
        bias_regularizer=regularizers.l2(1e-2),
    ),
    samplers=[mm.InBatchSampler()],
    embedding_options=mm.EmbeddingOptions(infer_embedding_sizes=True),
    metrics=[
        mm.RecallAt(1000),
    ],
    loss="categorical_crossentropy",
    logits_temperature=0.1,
)

batch_size = 3072

train_dl = tf_dataloader.BatchedDataset(
    Dataset(data_path + "/train/*.parquet", part_size="500MB", schema=schema),
    batch_size=batch_size,
)

test_dl = tf_dataloader.BatchedDataset(
    Dataset(data_path + "/test/*.parquet", part_size="500MB", schema=schema),
    batch_size=batch_size,
    shuffle=False,
)

from datetime import datetime

logdir = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")
tensorboard_callback = tf.keras.callbacks.TensorBoard(
    log_dir=logdir, update_freq=10, write_steps_per_second=True
)

model.set_retrieval_candidates_for_evaluation(train_dl.data)

opt = tf.keras.optimizers.Adam(learning_rate=0.001)
model.compile(opt, run_eagerly=False)

############ 1 - TRAINING ############

model.fit(
    train_dl,
    epochs=10,
    # steps_per_epoch=1,
    validation_data=test_dl,
    validation_steps=1,
    callbacks=[tensorboard_callback],
)

############ 2 - EVALUATION ############

eval_metrics = model.evaluate(test_dl, return_dict=True)
print("2 - EVALUATION METRICS: ", eval_metrics)

def numpy_recall(labels, top_item_ids, k):
    return np.equal(np.expand_dims(labels, -1), top_item_ids[:, :k]).max(axis=-1).mean()

############ 3 - CHECKING EVALUATION USING TOPK-RECOMMENDER ############

item_dataset = unique_rows_by_features(train_dl.data, Tags.ITEM, Tags.ITEM_ID)
model._load_or_update_candidates_index()
topk_recommender_model = model.to_top_k_recommender(Dataset(item_dataset.to_ddf()), k=1000)

""" BLOCK-01
topk_output = topk_recommender_model.predict(test_dl)
"""

""" BLOCK-02
topk_predictions, topk_items = topk_output
test_df = test_dl.data.to_ddf()
positive_item_ids = cupy.asnumpy(test_df["item_id"].compute().values)

recall_at_1000 = numpy_recall(positive_item_ids, topk_items, k=1000)
print(f"3 - TOP-K PREDICTION - RECALL@1000 = {recall_at_1000}")
"""

############ 4 - MANUALLY COMPUTING TOP-K PREDICTIONS ############

item_dataset = unique_rows_by_features(train_dl.data, Tags.ITEM, Tags.ITEM_ID)
candidates_dataset_df = IndexBlock.get_candidates_dataset(
    block=model.retrieval_block.item_block(), data=item_dataset, id_column="item_id"
)
item_tower_ids, item_tower_embeddings = IndexBlock.extract_ids_embeddings(
    candidates_dataset_df, check_unique_ids=True
)

recall_at_10_list = []
recall_at_1000_list = []
for batch in test_dl:
    batch_item_tower_embeddings = model.retrieval_block.item_block()(batch[0])
    batch_query_tower_embeddings = model.retrieval_block.query_block()(batch[0])
    positive_scores = tf.reduce_sum(
        tf.multiply(batch_item_tower_embeddings, batch_query_tower_embeddings), axis=-1
    )

    batch_user_scores_all_items = tf.matmul(
        batch_query_tower_embeddings, item_tower_embeddings, transpose_b=True
    )
    top_scores, top_indices = tf.math.top_k(batch_user_scores_all_items, k=1000)
    top_ids = tf.gather(item_tower_ids, top_indices)

    batch_pos_item_id = tf.squeeze(batch[0]["item_id"])
    recall_at_1000 = numpy_recall(batch_pos_item_id, top_ids, k=1000)

    recall_at_1000_list.append(recall_at_1000)

print(
    f"4 - MANUAL TOP-K PREDICTION - RECALL@1000 = {np.mean(recall_at_1000_list)}"
)
```
